### PR TITLE
usbutils: 014 (new formula)

### PR DESCRIPTION
**Pull request by nevack**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Following discussion here https://github.com/libusb/libusb/issues/1011
And this StackOverflow comment https://stackoverflow.com/questions/17058134/is-there-an-equivalent-of-lsusb-for-os-x#comment90807698_51579385

### CODE REVIEW

**Comment by nevack** [comment]
FYI @mikhailai

**Comment by SMillerDev** [comment]
Let's split out the two requests into different pull requests because I don't think deleting the other formula is the right approach.

**Comment by nevack** [comment]
> Let's split out the two requests into different pull requests because I don't think deleting the other formula is the right approach.

Okay.
Is it ok just to add new formula for `usbutils` and make it `conflicts with: lsusb` in this PR?
I'll drop `lsusb` deletion and `rename` commit.

**Comment by SMillerDev** [comment]
>  Is it ok just to add new formula for usbutils and make it conflicts with: lsusb in this PR?

Yeah, that sounds good to me

**Comment by nevack** [comment]
> > Is it ok just to add new formula for usbutils and make it conflicts with: lsusb in this PR?
> 
> Yeah, that sounds good to me

Done!

File: Formula/usbutils.rb
Comment (nevack) [inline Formula/usbutils.rb:17]: Not sure on this one

File: Formula/usbutils.rb
Comment (SMillerDev) [inline Formula/usbutils.rb:6]: We actually have this formula in homebrew.

File: Formula/usbutils.rb
Comment (SMillerDev) [inline Formula/usbutils.rb:42]:
```suggestion
    assert_empty shell_output("#{bin}/lsusb")
```

File: Formula/usbutils.rb
Comment (SMillerDev) [inline Formula/usbutils.rb:37]: Exiting with 1 is preferable to a version test that doesn't execute a real code path.

File: Formula/usbutils.rb
Comment (SMillerDev) [inline Formula/usbutils.rb:29]: Can't we disable them with configure?

File: Formula/usbutils.rb
Comment (SMillerDev) [inline Formula/usbutils.rb:26]:
```suggestion
    system "./configure", "--disable-debug",
                          *std_configure_args
```

File: Formula/usbutils.rb
Comment (SMillerDev) [inline Formula/usbutils.rb:22]: This shouldn't be needed

File: Formula/usbutils.rb
Comment (SMillerDev) [reply Formula/usbutils.rb:17]: Or seems like it depends on Linux if anything. Let's remove it.

File: Formula/usbutils.rb
Comment (nevack) [reply Formula/usbutils.rb:6]: I'll edit this comment.
The real issue is that macOS doesn't expose devices in `/dev` like it's in linux.
I think Homebrew `systemd` is not sufficient for this to work.

@tormodvolden, the maintainer of dfu-util, says it's possible to upstream patch to be able to build usbutils without `libudev`.

File: Formula/usbutils.rb
Comment (nevack) [reply Formula/usbutils.rb:22]: Why?

File: Formula/usbutils.rb
Comment (nevack) [reply Formula/usbutils.rb:29]: Nope, this linker flags are hardcoded. https://github.com/gregkh/usbutils/blob/master/Makefile.am#L10

File: Formula/usbutils.rb
Comment (nevack) [reply Formula/usbutils.rb:37]: Okay, I'll change it.

File: Formula/usbutils.rb
Comment (SMillerDev) [reply Formula/usbutils.rb:29]: I was more thinking of disabling the Linux utilities with that

File: Formula/usbutils.rb
Comment (SMillerDev) [reply Formula/usbutils.rb:22]: Because only in rare occasions you'd need to regenerate the configure script. In almost all formulae it works out of the box.

File: Formula/usbutils.rb
Comment (SMillerDev) [reply Formula/usbutils.rb:6]: Maybe we can make the formula Linux only at first then? And try and patch it upstream to avoid keeping an old version forever.

Comment (nevack) [reply Formula/usbutils.rb]: Nope, the whole point is to have real lsusb on macOS.

I think we can freeze at 007 on macOS and build latest tag on linux.

But I think it's more reasonable to make this formula macOS-only, as linux users usually have usbutils as part of installation.

File: Formula/usbutils.rb
Comment (nevack) [reply Formula/usbutils.rb:20]: There no configure script in archive downloaded from tag on GitHub.
That's why I added `autoreconf`

File: Formula/usbutils.rb
Comment (nevack) [reply Formula/usbutils.rb]: It's not empty, if there at least one USB device in system, changed tests, let's see how it works now.

File: Formula/usbutils.rb
Comment (SMillerDev) [reply Formula/usbutils.rb]: That's really at odds with https://docs.brew.sh/Acceptable-Formulae#niche-or-self-submitted-stuff though. We don't want to have old versions of software if we can avoid it.

Comment (nevack) [reply Formula/usbutils.rb]: It's still a way better to have a real `lsusb` than some wrapper shell script.

@tormodvolden has patch for compiling latest `usbutils` without `libudev`.
If not having the most recent version is the issue, I think, we can add the patch to this formula on macOS

File: Formula/usbutils.rb
Comment (nevack) [reply Formula/usbutils.rb:25]: Hmm, I'll check if there some flags about this.

**Comment by nevack** [comment]
Moved to latest version with patch for macOS

**Comment by SMillerDev** [comment]
Is that patch available somewhere online? Because we should submit it upstream and use that as the source if possible.

**Comment by tormodvolden** [comment]
The patch can be found online but is also verbatim included in this merge request. It is fine for just getting it to build and run on macOS, by disabling Linux-only stuff. It is however not yet generalized to be upstreamable, which would be quite more work.

What might even work better in the 007 version is the look-up of manufacturer and product strings. In any case, any of the old (but fully usable) 007 or the stripped 014 version is better than the systemprofiler joke, which is an abuse of the "lsusb" name. It is quite tiresome to ask macOS users to run lsusb for only to receive the non-compatible output from systemprofiler. I hope you can give another name to the systemprofiler wrapper.

File: Formula/usbutils.rb
Comment (tormodvolden) [reply Formula/usbutils.rb:23]: > There no configure script in archive downloaded from tag on GitHub. That's why I added `autoreconf`

usbutils releases are here https://mirrors.edge.kernel.org/pub/linux/utils/usb/usbutils/
github is just the git repo.

**Comment by nevack** [comment]
@SMillerDev Without autoreconf changes in `configure.ac` introduced by patch won't be picked up in `./configure`.

**Comment by nevack** [comment]
Fixed man pages installation

**Comment by nevack** [comment]
@SMillerDev Is there way to bypass `patch` audit check for new formula?

**Comment by SMillerDev** [comment]
I don't think there is, but I'll set it so that it builds everywhere regardless of that.

**Comment by nevack** [comment]
Rebased to trigger builds

**Comment by nevack** [comment]
Everything, other than audit for patch, is good.

**Comment by SMillerDev** [comment]
Can you submit the patch to upstream to at least start a discussion?

If you can't submit it upstream you can submit it to https://github.com/Homebrew/formula-patches and reference it in an issue upstream. Either way I would really want upstream to be aware of this patch before we make an exception for including it here.

**Comment by tormodvolden** [comment]
The patch was a quick hack to make lsusb build on Windows/MinGW, and it works on macOS also. However, it does 3 things:
- disable udev functionality (lookup of udev USB database for vendor and product strings)
- disable readlink() calls (not available on MinGW, used in sysfs lookup for string descriptors)
- disable POSIX signal handling (not available on MinGW, and not needed for lsusb)

Only the first would be necessary on macOS, I believe. It would be worth breaking it up into these parts, and further transform configure.ac libudev detection to set a HAVE_UDEV config flag instead of unconditionally require libudev, then substute #if 0 in the patch with #ifdef HAVE_UDEV where appropriate. Maybe upstream could accept this part (although they went linux-only after the 007 release). Longer term it would be nice to reinstate the /usr/share/misc/usb.ids lookup from 007, but not sure if upstream would like to maintain that.

**Comment by tormodvolden** [comment]
I have done many of the changes that I suggested above. Please try out https://github.com/tormodvolden/usbutils/tree/portable

I have tested it on Linux and compile-tested on Windows (MinGW). I have not made an upstream merge request yet, it doesn't make much sense before you have tested it on macOS.

**Comment by nevack** [comment]
@tormodvolden It still doesn't build on macOS with:
`ld: unknown option: --as-needed`

Need to remove this line https://github.com/tormodvolden/usbutils/blob/master/Makefile.am#L11 to make this work.

**Comment by nevack** [comment]
@SMillerDev Brew either complains about using fork repo or using a patch in new formula.

**Comment by tormodvolden** [comment]
> @SMillerDev Brew either complains about using fork repo or using a patch in new formula.

Um, I hadn't intented to host a patch set for homebrew there. It is just a staging repo, hopefully also for a future merge request. It might be rebased and force-updated too (I did yesterday).  Feel free to pick the patches though, if they work.

**Comment by SMillerDev** [comment]
The patch warning I'm okay waving when upstream is aware of the changes. But we can't use a fork.

**Comment by tormodvolden** [comment]
If it needs changes to build on macOS I would like to include them, of course. Indeed, I see now that I commented out the AM_LDFLAGS myself when I built with the original quick-and-dirty patch on macOS. It is not obvious to me what is the best way to make those flags being added only when ld is gcc.

**Comment by tormodvolden** [comment]
I added a commit which should take care of the --as-needed linker option.

**Comment by nevack** [comment]
> The patch warning I'm okay waving when upstream is aware of the changes. But we can't use a fork.

https://github.com/Homebrew/formula-patches/pull/408

**Comment by nevack** [comment]
I'll update patch link to https://github.com/Homebrew/formula-patches when https://github.com/Homebrew/formula-patches/pull/408 is merged.

**Comment by SMillerDev** [comment]
Can we use https://github.com/gregkh/usbutils/pull/134.patch? Or is that one larger than the formula-patches patch?

**Comment by nevack** [comment]
> Can we use https://github.com/gregkh/usbutils/pull/134.patch? Or is that one larger than the formula-patches patch?

They are the same, I just thought that it's required to import patches to https://github.com/Homebrew/formula-patches

**Comment by nevack** [comment]
> https://github.com/gregkh/usbutils/pull/134.patch

This type of url gives errors:
```
Use a commit hash URL rather than an unstable pull request URL: https://github.com/gregkh/usbutils/pull/134.patch
```
and
```
Use a commit hash URL rather than patch-diff: https://patch-diff.githubusercontent.com/raw/gregkh/usbutils/pull/134.patch
```

**Comment by nevack** [comment]
@SMillerDev How should I include patch from https://github.com/gregkh/usbutils/pull/134 ?

**Comment by SMillerDev** [comment]
Since it's not a single commit, you can submit this minimal version to formula-patches with a comment in the formula linking to the PR.

**Comment by nevack** [comment]
> Since it's not a single commit, you can submit this minimal version to formula-patches with a comment in the formula linking to the PR.

It's already there, I cannot update formula until it's merged.
https://github.com/Homebrew/formula-patches/pull/408

**Comment by nevack** [comment]
@SMillerDev What else can be done to help this PR get accepted?

**Comment by tormodvolden** [comment]
AFAICS all mentioned concerns have been addressed now, and someone with write access just needs to push the Approve button.

File: Formula/usbutils.rb
Comment (SMillerDev) [inline Formula/usbutils.rb:2]:
```suggestion
  desc "Get detailed info about USB devices"
```

File: Formula/usbutils.rb
Comment (SMillerDev) [inline Formula/usbutils.rb:15]:
```suggestion
```

File: Formula/lsusb.rb
Comment (SMillerDev) [inline Formula/lsusb.rb:21]:
```suggestion
  conflicts_with "usbutils", because: "both provide a `lsusb` binary"
```

File: Formula/usbutils.rb
Comment (tormodvolden) [reply Formula/usbutils.rb:2]: or "List"

Comment (SMillerDev) [reply Formula/usbutils.rb:2]: Yeah, this is also an option
```suggestion
  desc "List detailed info about USB devices"
```

File: Formula/usbutils.rb
Comment (tormodvolden) [inline Formula/usbutils.rb:22]: This should point to the official formula-patches location, right?

File: Formula/usbutils.rb
Comment (tormodvolden) [inline Formula/usbutils.rb:5]: Shouldn't the official upstream release be used instead of a git snapshot?

Comment (nevack) [reply Formula/usbutils.rb:5]: @SMillerDev can you help with this?

Comment (tormodvolden) [reply Formula/usbutils.rb:5]: See https://github.com/Homebrew/homebrew-core/pull/89341#discussion_r750726447. autoreconf wouldn't be needed.

**Comment by tormodvolden** [comment]
@nevack that's very interesting, but please one thing at a time. Please get this into brew first :)

File: Formula/usbutils.rb
Comment (nevack) [reply Formula/usbutils.rb:5]: `configure` caches `libudev` requirement without `autoreconf`:
```
configure: error: Package requirements (libudev >= 196) were not met:

No package 'libudev' found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.
```

File: Formula/usbutils.rb
Comment (nevack) [reply Formula/usbutils.rb]: Removed

File: Formula/lsusb.rb
Comment (nevack) [reply Formula/lsusb.rb]: Fixed

**Comment by nevack** [comment]
@SMillerDev Applied suggestions.

`autoreconf` is still a thing, unfortunately.

I will update patch url when https://github.com/Homebrew/formula-patches/pull/408 is merged

File: Formula/usbutils.rb
Comment (tormodvolden) [reply Formula/usbutils.rb]: Yes, you're right of course, the autoreconf is still needed. In one of my patches configure.ac is changed to make udev optional. For this to be reflected in configure, autoreconf must be run to regenerate it.

File: Formula/lsusb.rb
Comment (nevack) [inline Formula/lsusb.rb:19]: This was a mistake, I’ll fix it

**Review by carlocab — COMMENTED** [review]
Forgot to submit my comments I wrote much earlier...

File: Formula/lsusb.rb
Comment (tormodvolden) [reply Formula/lsusb.rb:19]: It is less useful on Linux where everybody will have access to usbutils through normal channels. But it will build on all platforms.

File: Formula/usbutils.rb
Comment (tormodvolden) [reply Formula/usbutils.rb:39]: Good idea. It will be more reproducible, with no output and return code 1.
```
$ lsusb -d ffff:ffff
$ echo $?
1
```

File: Formula/lsusb.rb
Comment (carlocab) [reply Formula/lsusb.rb:19]: It seems to rely on `system_profiler`, though, which is macOS-only AFAIK.

Comment (nevack) [reply Formula/lsusb.rb:19]: system_profiler is macOS specific.
This change was unintentional, I’ll revert it

Comment (tormodvolden) [reply Formula/lsusb.rb:19]: Now you're mixing up things. The old "lsusb" package was a wrapper for system_profiler. This package "usbutils" delivers the real lsusb.

File: Formula/usbutils.rb
Comment (carlocab) [reply Formula/usbutils.rb:39]:
```ruby
assert_empty shell_output("#{bin}/lsusb -d ffff:ffff", 1)
```

File: Formula/lsusb.rb
Comment (carlocab) [reply Formula/lsusb.rb:19]: > Now you're mixing up things. The old "lsusb" package was a wrapper for system_profiler. This package "usbutils" delivers the real lsusb.

Nope, this line is in the old "`lsusb`" wrapper for `system_profiler`.

Comment (tormodvolden) [reply Formula/lsusb.rb:19]: You might be confused by the GitHub Actions for some reason running the old lsusb wrapper.

Comment (carlocab) [reply Formula/lsusb.rb:19]: There are two files being modified in this PR: `Formula/lsusb.rb`, which is the wrapper for `system_profiler`, and `Formula/usbutils.rb`, which is the real `lsusb`.

This comment refers to a line inside `Formula/lsusb.rb`.

Comment (tormodvolden) [reply Formula/lsusb.rb:19]: I was confused and mixing things up! I didn't notice this PR also modifies lsusb.rb at the same time.

Comment (tormodvolden) [reply Formula/lsusb.rb:19]: Can you also change the description of lsusb.rb from "lsusb command for Mac OS X" to "lsusb-like wrapper around system_profiler"?

Comment (carlocab) [reply Formula/lsusb.rb:19]: It seems it made you confused enough to be confused about my being confused 🙂

File: Formula/usbutils.rb
Comment (tormodvolden) [reply Formula/usbutils.rb:39]: (Of course, lsusb --version is also pretty reproducible but it exercises less of the code.)

**Comment by nevack** [comment]
@carlocab Resolved discussions

**Comment by nevack** [comment]
Updated patch URL

**Review by carlocab — DISMISSED** [review]

**Comment by nevack** [comment]
@carlocab Is there a way to bypass this check?

```
usbutils:
    * Formulae should not require patches to build. Patches should be submitted and accepted upstream first.
```

**Comment by carlocab** [comment]
No, but we can merge despite failing CI. (For now. That'll probably change in this repo at some point.)

**Review by carlocab — APPROVED** [review]
Split up your commit for you. Feel free to ping me when CI is done.

**Pull request closed without merging**
